timestamp-impl/src/test/java/com/palantir/timestamp/PersistentTimestampServiceTest.java

```diff
@@ -71,6 +73,52 @@ public void testFastForward() {
     }
 
     @Test
+    public void incrementUpperLimitIfOneMinuteElapsedSinceLastUpdate() throws InterruptedException {
+        Clock clock = mock(Clock.class);
+        when(clock.getTimeMillis()).thenReturn(0L, TWO_MINUTES_IN_MILLIS, 2 * TWO_MINUTES_IN_MILLIS, 3 * TWO_MINUTES_IN_MILLIS);
+        TimestampBoundStore timestampBoundStore = initialTimestampBoundStore();
+        PersistentTimestampService persistentTimestampService = PersistentTimestampService.create(timestampBoundStore, clock);
+
+        persistentTimestampService.getFreshTimestamp();
+        Thread.sleep(10);
+        persistentTimestampService.getFreshTimestamp();
+        Thread.sleep(10);
+        verify(timestampBoundStore, times(2)).storeUpperLimit(anyLong());
+    }
+
+    @Test
+    public void incrementUpperLimitOnFirstFreshTimestampRequest() {
+        TimestampBoundStore timestampBoundStore = initialTimestampBoundStore();
+        PersistentTimestampService persistentTimestampService = PersistentTimestampService.create(timestampBoundStore);
+
+        persistentTimestampService.getFreshTimestamp();
+
+        verify(timestampBoundStore).storeUpperLimit(PersistentTimestampService.ALLOCATION_BUFFER_SIZE);
+    }
+
+    @Test
+    public void multipleFreshTimestampRequestsShouldIncreaseUpperLimitOnlyOnce() {
+        TimestampBoundStore timestampBoundStore = initialTimestampBoundStore();
+        PersistentTimestampService persistentTimestampService = PersistentTimestampService.create(timestampBoundStore);
+
+        getFreshTimestampsInParallel(persistentTimestampService, 20);
+
+        verify(timestampBoundStore, times(1)).storeUpperLimit(PersistentTimestampService.ALLOCATION_BUFFER_SIZE);
+    }
+
+    @Test
```

Comment:
But that's not what the test is testing. I think this is going very inside out, we're writing an implementation (in code or in our heads) then writing tests to check the different oddities that code results in.

Ideally we want to be figuring out the behaviour we want (in tests, or otherwise) *then* figuring out the code.

If an important part of this class is that we only load timestamps when they're requested, then that's a good test (that we have). However, I don't think we want what this test is checking, even if our current implementation makes it true and we like our current implementation.